File: js/encryption/hybridEncryption.js
// hybridEncryption.js

const crypto = require('crypto');

// Generate RSA keys
const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
  modulusLength: 2048,
});

// Symmetric encryption with AES
const algorithm = 'aes-256-cbc';
const key = crypto.randomBytes(32);
const iv = crypto.randomBytes(16);

// Encrypt function
function encrypt(text) {
  // Symmetrically encrypt the message
  const cipher = crypto.createCipheriv(algorithm, key, iv);
  let encryptedMessage = cipher.update(text, 'utf8', 'hex');
  encryptedMessage += cipher.final('hex');

  // Asymmetrically encrypt the AES key
  const encryptedKey = crypto.publicEncrypt(publicKey, key);

  return {
    encryptedMessage,
    encryptedKey: encryptedKey.toString('hex'),
    iv: iv.toString('hex'),
  };
}

// Decrypt function
function decrypt(encryptedMessage, encryptedKeyHex, ivHex) {
  const iv = Buffer.from(ivHex, 'hex');
  const encryptedKey = Buffer.from(encryptedKeyHex, 'hex');

  // Asymmetrically decrypt the AES key
  const key = crypto.privateDecrypt(privateKey, encryptedKey);

  // Symmetrically decrypt the message
  const decipher = crypto.createDecipheriv(algorithm, key, iv);
  let decrypted = decipher.update(encryptedMessage, 'hex', 'utf8');
  decrypted += decipher.final('utf8');

  return decrypted;
}

module.exports = { encrypt, decrypt };

File: python/encryption/hybrid_encryption.py
# hybrid_encryption.py

import os
from cryptography.hazmat.primitives.asymmetric import rsa, padding
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives import serialization

# Generate RSA key pair
private_key = rsa.generate_private_key(public_exponent=65537, key_size=2048)
public_key = private_key.public_key()

# Encrypt function
def encrypt(message: str) -> dict:
    # Symmetric encryption (AES)
    aes_key = os.urandom(32)
    iv = os.urandom(16)
    cipher = Cipher(algorithms.AES(aes_key), modes.CFB(iv))
    encryptor = cipher.encryptor()
    encrypted_message = encryptor.update(message.encode()) + encryptor.finalize()

    # Encrypt AES key with RSA
    encrypted_key = public_key.encrypt(
        aes_key,
        padding.OAEP(
            mgf=padding.MGF1(algorithm=hashes.SHA256()),
            algorithm=hashes.SHA256(),
            label=None
        )
    )
    
    return {
        'encrypted_message': encrypted_message,
        'encrypted_key': encrypted_key,
        'iv': iv
    }

# Decrypt function
def decrypt(encrypted_data: dict) -> str:
    encrypted_message = encrypted_data['encrypted_message']
    encrypted_key = encrypted_data['encrypted_key']
    iv = encrypted_data['iv']
    
    # Decrypt AES key with RSA
    aes_key = private_key.decrypt(
        encrypted_key,
        padding.OAEP(
            mgf=padding.MGF1(algorithm=hashes.SHA256()),
            algorithm=hashes.SHA256(),
            label=None
        )
    )

    # Symmetric decryption
    cipher = Cipher(algorithms.AES(aes_key), modes.CFB(iv))
    decryptor = cipher.decryptor()
    decrypted_message = decryptor.update(encrypted_message) + decryptor.finalize()
    
    return decrypted_message.decode()

if __name__ == "__main__":
    msg = "Hello, World!"
    encrypted_data = encrypt(msg)
    print(f"Encrypted: {encrypted_data}")
    print(f"Decrypted: {decrypt(encrypted_data)}")
